blog/2024-06-14-using-table-component.md

```diff
@@ -0,0 +1,486 @@
+                    _header_cell("Actions", "cog"),
+                ),
+            ),
+            rx.table.body(rx.foreach(State.users, show_customer)),
+            on_mount=State.load_entries,
+        )
+
+
+app = rx.App(
+    theme=rx.theme(appearance="light", has_background=True, radius="large", accent_color="grass"),
+    )
+app.add_page(main_table)
+```
+
+
+
+
+
+## Adding a New Customer 
+
+Next let's explore how to add a user to our database and therefore to our table. We have a function `add_customer_button`, which is a UI element that is rendered in the `main_table` function defined in the section above. 
+
+In this `add_customer_button` function there is an `rx.dialog` which opens when the `rx.button` "Add New Customer" is pressed. When the dialog is opened there is an `rx.form.root` with several `form_field` functions called inside. The `form_field` function returns an `rx.form.field`. It takes in several arguments, but most importantly takes in a `name` argument. The `name` prop is needed to submit with its owning form as part of a `name/value` pair. The content typed into the `rx.input` is submitted as the `value` in that `name/value` pair when the `rx.form` is submitted. The same is done for the `rx.radio` component, where the `name` is set and the radio item selected is passed as the `value` when the `rx.form` is submitted.
+
+
+
+To submit the `rx.form` with all the `name/value` pairs added, the "Submit Customer" button is pressed which is inside of an `rx.dialog.close` and an `rx.form.submit`, so it closes the dialog and it submits the form. When the form is submitted it runs the code `on_submit=State.add_customer_to_db,` running the event handler `add_customer_to_db`. This event handler sets the `State` var `current_user` to equal the form data just submitted in the `rx.form`. This event handler also does a database query to add the current `self.current_user`, which we just set to be the form response, to the database. Before this a check is done to see if the `email` the user has added is already an `email` in the database, and if it is then an `rx.window_alert` is thrown. 
+
+Finally once the new user is added to the database, the `load_entries` event handler is run again which loads all the data from the database and fills the table.
+
+
+```python
+class State(rx.State):
+    """The app state."""
+
+    users: list[Customer] = []
+    current_user: Customer = Customer()
+
+    ...
+
+    def add_customer_to_db(self, form_data: dict):
+        """Add a customer to the database."""
+        self.current_user = form_data
+        self.current_user["date"] = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
+        with rx.session() as session:
+            if session.exec(
+                select(Customer).where(Customer.email == self.current_user["email"])
+            ).first():
+                return rx.window_alert("User with this email already exists")
+            session.add(Customer(**self.current_user))
```

Comment:
```suggestion
            self.current_user = Customer(**form_data, date=datetime.now())
            session.add(Customer(self.current_user))
```

Reply:
this has already been added a few lines above here:
```
self.current_user["date"] = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
```

Reply:
the issue i have is that `self.current_user` should be a `Customer`, not a `dict`. As the code is written, it's treating `self.current_user` as a dict, which might be confusing to other event handlers that may want to access `self.current_user`